manifest: trusted-firmware-m: update to 2.1.0

tomi-font opened the pull request:
Update TF-M to version 2.1.0.

tomi-font committed:
manifest: trusted-firmware-m: update to 2.1.0

Update TF-M to version 2.1.0.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modules: tf-m: disable tf-m-tests revision checks

Permanently disable the revision checks made in the tf-m-tests
repository that were introduced with TF-M 2.1.0.

They fail because the expected upstream tags are not found in Zephyr's
tf-m-tests.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font commented:
`psa-arch-tests` cannot currently be updated from 23.06 to 24.03 as some tests start to fail.
Upstream projects need to incorporate fixes for this update to be possible.

frkv reviewed (approved):
LGTM (pending passing tests and updates to correct SHA)

tomi-font committed:
Revert "tf-m: Silence harmless version warning"

This reverts commit 33786b55abc783cfed3a80dd0197eca26730baef.

Reverting it because the upstream PR has been abandoned, and
this extra CMake variable is causing the warning
"Manually-specified variables were not used by the project".

As of now the version warning doesn't come up.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
samples: tfm_psa_test: fix CMake logs

Remove TF-M version number and explicit mention of the available test
suites; one was missing.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modules: tf-m: fix unused CMake variable warning

Define and pass down PSA_ARCH_TESTS_PATH only in the sample that
needs it (tfm_psa_test).
Otherwise, it provokes a CMake warning because the variable does not
get used in TF-M.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modules: tf-m: restore check on CONFIG_TFM_PARTITION_INITIAL_ATTESTATION

Restore the check that was introduced in
cd8d4ccad55544b38637632e747a3b00a8c01016 and removed in
cac7f4058ff3668b67fdb2c972ee9bb2fee1bfb5, rather than checking only
CONFIG_TFM_PSA_TEST_INITIAL_ATTESTATION in the tfm_psa_test sample.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modules: tf-m: support MCUboot signature types other than RSA-3072

With TF-M updated to 2.1.0 it now makes the signature type default
to EC-P256 for the mps2/an521/cpu0/ns board.
So far Zephyr had only supported and assumed that it was RSA-3072.

This brings support for other signature types, and changes the global
default to EC-P256.
The switch from RSA-3072 to EC-P256 reduces the flash usage by ~3.3KB
while having a negligible impact on RAM usage (increase of ~70 bytes)
when compiling the tfm_psa_test sample on mps2/an521/cpu0/ns and
nrf9160dk/nrf9160/ns without explicit optimizations.

The TFM_KEY_FILE_{S,NS} Kconfig options are moved inside an
`if TFM_BL2` as they are only used if MCUboot is included in TF-M.

The TF-M CMake variables MCUBOOT_KEY_{S,NS} are now set so that it's
possible to use signing keys located elsewhere than the default
location.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

frkv commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -258,6 +237,28 @@ config TFM_CONNECTION_BASED_SERVICE_API
+	string "The signature type used to sign the secure and non-secure firmware images."
+	default "EC-P256"
+	help
+	  Available types: RSA-2048, RSA-3072, EC-P256, EC-P384.
```
Should ed25519 be on this list as well?

tomi-font replied:
No: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/105/files#diff-c8402d394cb78c69fcf9bc110193858d9c6beb04277528dc74e30bcfe7a72e5bR50

ceolin commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -389,6 +401,8 @@ if (CONFIG_BUILD_WITH_TFM)
       set(pad_args --pad --pad-header)
     endif()
+    string(REPLACE "\${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}" "${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}"
+                    CONFIG_TFM_KEY_FILE_${SUFFIX} ${CONFIG_TFM_KEY_FILE_${SUFFIX}})
```
Are `CONFIG_TFM_KEY_FILE_S` and `CONFIG_TFM_KEY_FILE_NS` valid symbols ? Isn't TFM using `TFM_KEY_S` and `TFM_KEY_NS` ?

ceolin commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -35,6 +35,18 @@ if (CONFIG_BUILD_WITH_TFM)
+    elseif ("${CONFIG_TFM_MCUBOOT_SIGNATURE_TYPE}" MATCHES "^EC-P")
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_USE_PSA_CRYPTO=ON)
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_HW_KEY=OFF)
+      list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_BUILTIN_KEY=ON)
```
Why these options based on the type of key ? the only dependency that is clear to me is `MCUBOOT_BUILTIN_KEY` requires `MCUBOOT_USE_PSA_CRYPTO` but I don't understand why, for example, you set HW_KEY for RSA and BUILTIN_KEY for EC-P.

Maybe just a comment explaining it would make things easier.

tomi-font replied:
I added a comment to explain why.
RSA without `MCUBOOT_HW_KEY` and EC-P without `MCUBOOT_BUILTIN_KEY` failed to boot (`Image in the primary slot is not valid!` and `Provisioning failed`, respectively).

ceolin commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -258,6 +237,28 @@ config TFM_CONNECTION_BASED_SERVICE_API
+	help
+	  The path and filename for the .pem file containing the private key
+	  that should be used by the BL2 bootloader when signing non-secure
+	  firmware images.
```
Are these symbols still used in TF-M or they were replaced ?

tomi-font committed:
samples: psa_crypto: fix setting of TF-M's CRYPTO_ENGINE_BUF_SIZE

It used to be set as a CMake variable, but TF-M's build system does
not look at such a variable.
Instead, define an additional configuration header file that defines
CRYPTO_ENGINE_BUF_SIZE, and pass it to TF-M's build system.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
boards: st: b_u585_iot02a: stick to RSA signature type in TF-M's MCUboot

The default has been changed to EC-P256, but there are issues with
this board when TF-M is compiled with MCUBOOT_USE_PSA_CRYPTO (required
by EC-P256) because of its integration in TF-M.
Do not enable PSA crypto on this board for the time being.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

mswarowsky reviewed (approved):
lgtm

frkv reviewed (approved):
LGTM

ithinuel commented on samples/tfm_integration/psa_crypto/src/main.c:
```diff
@@ -44,5 +44,7 @@ int main(void)
 	/* Dump any queued log messages, and wait for system events. */
 	al_dump_log();
 
+	LOG_INF("Done.");
+
```
Is this change required?

tomi-font replied:
To this and your other comment: Not strictly required as in the sample would be broken without them, but those very much help someone looking at the logs to understand that the sample is done instead of having hung. So to me yes, those are important enough to have.

ithinuel commented on samples/tfm_integration/psa_crypto/src/psa_crypto.c:
```diff
@@ -510,7 +510,7 @@ void crp_generate_csr(void)
 	al_dump_log();
 
-	printf("%s\n", json_encoded_buf);
+	printf("%s\n\n", json_encoded_buf);
```
Ditto
